### PR TITLE
Introduce custom TLV support for upstream PP2 headers

### PR DESCRIPTION
**Pull request by timflannagan**
Commit Message:

This commit adds support for injecting custom TLVs into the Proxy Protocol v2 (PP2) header for upstream transport sockets, providing xDS control planes with greater flexibility in building upstream PP2 headers.

Previously, upstream PP2 headers were limited to TLVs passthrough'd from the downstream connection when using the Proxy Protocol listener. This approach constrained customization options for upstream connections.

With this change, users can now define custom TLVs by specifying host metadata in a well-known namespace, enabling dynamic and granular control over the PP2 header content. For example:

```yaml
clusters:
      - name: httpbin
        load_assignment:
          ...
          endpoints:
          - lbEndpoints:
            - metadata:
                filter_metadata:
                  envoy.transport_socket_match:
                    outbound-proxy: true
                typed_filter_metadata:
                  envoy.transport_sockets.proxy_protocol:
                    "@type": type.googleapis.com/envoy.extensions.transport_sockets.proxy_protocol.v3.CustomTlvMetadata
                    entries:
                      - type: 0x96
                        value: Zm9v # foo
                      - type: 0x97
                        value: YmFy # bar
               ...

```

By decoupling upstream PP2 header customization from the downstream listener configuration, this change unlocks more flexible use cases for Proxy Protocol in upstream connections.

Earlier implementations considered extending the `upstream_proxy_protocol` protobuf definition to allow users to configure an array of TLV key-value pairs. However, this approach was deemed suboptimal as it shifted additional complexity to control planes responsible for generating xDS configuration.

Similarly, initial attempts looked into whether we could re-use the `envoy.network.proxy_protocol_options` filter metadata namespace and read from that filter metadata much later in the request function chain. While possible, it would require more refactoring, increase risk, etc. that I was comfortable with at the time.

Additional Description: N/A
Risk Level: Low
Testing: Unit & Integration
Docs Changes: TODO
Release Notes: TODO
Platform Specific Features: N/A.
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/common/proxy_protocol/proxy_protocol_header.cc [hunk 1]
```diff
@@ -111,12 +111,26 @@ void generateV2Header(const Network::Address::Ip& source_address,
 }
 
 bool generateV2Header(const Network::ProxyProtocolData& proxy_proto_data, Buffer::Instance& out,
-                      bool pass_all_tlvs, const absl::flat_hash_set<uint8_t>& pass_through_tlvs) {
-  uint64_t extension_length = 0;
-  for (auto&& tlv : proxy_proto_data.tlv_vector_) {
```
Comment (timflannagan) [inline source/extensions/common/proxy_protocol/proxy_protocol_header.cc:116]: Note: I initially tried expanding the ProxyProtocolData structure to house the custom TLVs, but I ran into several issues trying to modify or inject TLV data after initial construction. I think there's a way to refactor this s.t. we don't have to DI the custom TLVs in this common generateV2Header function, but this implementation was dramatically easier and I didn't think it was worth chasing a refactor at this time.

File: source/extensions/common/proxy_protocol/proxy_protocol_header.cc [hunk 2]
```diff
@@ -111,12 +111,26 @@ void generateV2Header(const Network::Address::Ip& source_address,
...
+                      bool pass_all_tlvs, const absl::flat_hash_set<uint8_t>& pass_through_tlvs,
+                      const std::unordered_map<uint8_t, std::vector<unsigned char>>& custom_tlvs) {
```
Comment (timflannagan) [inline source/extensions/common/proxy_protocol/proxy_protocol_header.cc]: I decided to make the custom_tlvs parameter a map instead of vector -- largely to prevent duplicate TLV keys between the passthrough & custom TLV keyset -- for quick O(1) key lookups.

File: test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc [hunk 3]
```diff
@@ -463,6 +465,27 @@ class ProxyProtocolTLVsIntegrationTest : public testing::TestWithParam<Network::
...
+        ProtobufWkt::Struct metadata_struct;
+        auto* fields = metadata_struct.mutable_fields();
+        for (const auto& [type, value] : custom_tlvs_) {
+            ProtobufWkt::Value metadata_value;
+            metadata_value.set_string_value(value);
+            (*fields)[absl::StrFormat("0x%02X", type)] = metadata_value;
+        }
+        metadata->mutable_filter_metadata()->insert(
+        Protobuf::MapPair<std::string, ProtobufWkt::Struct>(
+            "envoy.transport_sockets.proxy_protocol", metadata_struct)); // FIXME: avoid hardcoding.
```
Comment (timflannagan) [inline test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc]: Still relevant. I need to figure out how to add a constant variable for this metadata namespace.

File: test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc [hunk 4]
```diff
@@ -681,5 +705,55 @@ TEST_P(ProxyProtocolTLVsIntegrationTest, TestV2ProxyProtocolPassWithTypeLocal) {
   ASSERT_TRUE(fake_upstream_connection_->waitForDisconnect());
 }
 
+TEST_P(ProxyProtocolTLVsIntegrationTest, TestV2TLVProxyProtocolWithCustomMetadata) {
```
Comment (timflannagan) [inline test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc:712]: We might also want an integration test for 1.) passthrough + custom TLVs, or 2.) no downstream PP2 listener/network filter, but upstream transport socket is PP2.

File: test/integration/fake_upstream.cc [hunk 5]
```diff
@@ -1037,7 +1037,7 @@ AssertionResult FakeRawConnection::waitForData(uint64_t num_bytes, std::string*
   ENVOY_LOG(debug, "waiting for {} bytes of data", num_bytes);
   if (!time_system_.waitFor(lock_, absl::Condition(&reached), timeout)) {
     return AssertionFailure() << fmt::format(
-               "Timed out waiting for data. Got '{}', waiting for {} bytes.", data_, num_bytes);
+               "Timed out waiting for data. Got '{}', length {}, waiting for {} bytes.", data_, data_.size(), num_bytes);
```
Comment (timflannagan) [inline test/integration/fake_upstream.cc]: This is debug cruft. It can be removed.

File: test/mocks/stream_info/mocks.h [hunk 6]
```diff
@@ -192,6 +192,7 @@ class MockStreamInfo : public StreamInfo {
   uint64_t bytes_received_{};
   uint64_t bytes_sent_{};
   std::shared_ptr<Network::ConnectionInfoSetterImpl> downstream_connection_info_provider_;
+  Upstream::HostDescriptionConstSharedPtr upstream_host_;
```
Comment (timflannagan) [inline test/mocks/stream_info/mocks.h]: AFAIK, this is cruft from a previous implementation and can be removed.

File: source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc [hunk 7]
```diff
@@ -165,6 +168,69 @@ void UpstreamProxyProtocolSocketFactory::hashKey(
   }
 }
 
+std::unordered_map<uint8_t, std::vector<unsigned char>> UpstreamProxyProtocolSocket::buildCustomTLVs() {
```
Comment (timflannagan) [inline source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc]: Note, I previously tried populating this data in the UpstreamProxyProtocolSocket constructor, but the issue I was seeing with that approach is the transport socket callbacks haven't been populated until after construction, and therefore we can't read from the filter metadata namespace at that point in the function chain.

File: source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc [hunk 8]
```diff
@@ -165,6 +168,69 @@ void UpstreamProxyProtocolSocketFactory::hashKey(
...
+  // Construct the map from the metadata.
+  const auto& metadata_struct = filter_it->second;
+  for (const auto& field : metadata_struct.fields()) {
+    const std::string& key = field.first;
+    const auto& value = field.second;
+
+    // Assume the metadata key is the TLV type in hex format (e.g., "0xD3").
+    uint8_t tlv_type;
+    try {
+      tlv_type = static_cast<uint8_t>(std::stoi(key, nullptr, 16));
```
Comment (yuval-k) [inline source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc]: WDYT abot using typed_metadata so we don't have to do this parsing?

File: source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc [hunk 9]
```diff
@@ -165,6 +168,69 @@ void UpstreamProxyProtocolSocketFactory::hashKey(
...
+    // Only process string values.
+    if (value.kind_case() != ProtobufWkt::Value::kStringValue) {
+      continue;
+    }
+
+    const std::string& string_value = value.string_value();
+    if (string_value.empty()) {
+      continue;
+    }
+    std::string sanitised_value = MessageUtil::sanitizeUtf8String(string_value);
```
Comment (yuval-k) [inline source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc]: why do we need to sanitize this? i think TLV can be bytes?

File: test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc (hunk 3 above)
Comment (yuval-k) [reply test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc]: see example: https://github.com/envoyproxy/envoy/blob/a4c2d71a18eccd981a877aa247a6393cbff601d8/source/common/upstream/upstream_impl.cc#L549

File: source/extensions/common/proxy_protocol/proxy_protocol_header.cc (hunk 2 above)
Comment (yuval-k) [reply source/extensions/common/proxy_protocol/proxy_protocol_header.cc]: these small scale vector perf is probably better

File: source/extensions/common/proxy_protocol/proxy_protocol_header.cc [hunk 10]
```diff
@@ -111,12 +111,26 @@ void generateV2Header(const Network::Address::Ip& source_address,
...
+  std::vector<Envoy::Network::ProxyProtocolTLV> combined_tlv_vector_;
+  for (const auto& [type, value] : custom_tlvs) {
+    combined_tlv_vector_.push_back({type, {value.begin(), value.end()}});
```
Comment (yuval-k) [inline source/extensions/common/proxy_protocol/proxy_protocol_header.cc:118]: should we do the check if already exists here too?

File: source/extensions/common/proxy_protocol/proxy_protocol_header.cc (hunk 2 above)
Comment (timflannagan) [reply source/extensions/common/proxy_protocol/proxy_protocol_header.cc]: Definitely agree perf would be better with a small vector here. I ran into several issues with previous implementations that tried to define a vector. Mainly because I wanted to re-use the ProxyProtocolTLV type:

```cpp
struct ProxyProtocolTLV {
  const uint8_t type;
  const std::vector<unsigned char> value;
};

using ProxyProtocolTLVVector = std::vector<ProxyProtocolTLV>;
```

IIRC, builds were failing and complaining that copy assignment was disabled for that type. I tried to evade that issue several times by using std::move, emplace, passing the parameter as a const ref, etc. but I wasn't able to get builds back online. Any idea on what I'd need to do there?

File: source/extensions/common/proxy_protocol/proxy_protocol_header.cc (hunk 10 above)
Comment (timflannagan) [reply source/extensions/common/proxy_protocol/proxy_protocol_header.cc:118]: Checking whether the custom TLV key already exists in the combined_tlv_vector_ variable? That should be empty after initialization on line 116.

File: source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc (hunk 9 above)
Comment (timflannagan) [reply source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc]: I'm blanking on why I added this now. FWIW, I realize too there's no negative unit tests for what we consider "invalid" string content. I'd have to go through the PP2 protocol documentation again to see whether there's any assumptions we need to capture in this code.

File: test/mocks/stream_info/mocks.h (hunk 6 above)
Comment (timflannagan) [reply test/mocks/stream_info/mocks.h]: Addressed in https://github.com/timflannagan/envoy/pull/1/commits/48c2f7ab22d146fd48fd800ab925a96e2d1c3453.

File: test/integration/fake_upstream.cc (hunk 5 above)
Comment (timflannagan) [reply test/integration/fake_upstream.cc]: https://github.com/timflannagan/envoy/pull/1/commits/48c2f7ab22d146fd48fd800ab925a96e2d1c3453

File: test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc (hunk 3 above)
Comment (timflannagan) [reply test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc]: Addressed in https://github.com/timflannagan/envoy/pull/1/commits/48c2f7ab22d146fd48fd800ab925a96e2d1c3453. Let me know about constant variable naming.

File: source/extensions/common/proxy_protocol/proxy_protocol_header.cc (hunk 1 above)
Comment (timflannagan) [reply source/extensions/common/proxy_protocol/proxy_protocol_header.cc:116]: The more I look at this code, I'm wondering whether we could re-use the existing PP listener filter metadata namespace to meet our use case. We just need to populate the PRoxyProtocolData structure later, e.g. in the PP transport socket implementation, in order to do this. I'm not sure what's best here.

**Comment by timflannagan** [comment]
TODO: Add docs for this metadata namespace?

File: test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc (hunk 4 above)
Comment (timflannagan) [reply test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc:712]: Added an integration test for passthrough + custom TLVs in https://github.com/timflannagan/envoy/pull/1/commits/974bc574b4334048e39d408716e5aa4352205e83.

File: source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc (hunk 8 above)
Comment (timflannagan) [reply source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc]: Addressed in https://github.com/timflannagan/envoy/pull/1/commits/fe77f985401b45fdd9b349b8f0dd73c5a9c5115d. I ended up embedding that type in the PP2 upstream transport socket proto, but it's a bit awkwardly placed since 1.) you can't set this config field in the upstream_proxy_protocol transport socket config, and 2.) I'm not sure we want to support populating custom TLVs from filter metadata and reading from the transport socket config.

File: source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc (hunk 9 above)
Comment (timflannagan) [reply source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc]: Resolving -- I don't think this is relevant anymore now that the implementation was refactored to use a typed metadata proto.

File: api/envoy/extensions/transport_sockets/proxy_protocol/v3/upstream_proxy_protocol.proto [hunk 11]
```diff
@@ -17,6 +17,19 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
+  repeated TlvEntry entries = 1;
+}
+
+message TlvEntry {
+  // type is the type of the TLV. The PP2 protocol expects this type to be a uint8_t.
+  uint32 type = 1 [(validate.rules).uint32 = {lt: 256}];
+  // value is the value of the TLV.
+  //
+  // TODO(tim): should this be a bytes field instead?
+  string value = 2 [(validate.rules).string = {min_len: 1}];
```
Comment (timflannagan) [inline api/envoy/extensions/transport_sockets/proxy_protocol/v3/upstream_proxy_protocol.proto]: Still an open question after re-reading the protocol documentation:

```
The information in those bytes will be arranged in Type-Length-Value (TLV
vectors) in the following format.  The first byte is the Type of the vector.
The second two bytes represent the length in bytes of the value (not included
the Type and Length bytes), and following the length field is the number of
bytes specified by the length.

        struct pp2_tlv {
            uint8_t type;
            uint8_t length_hi;
            uint8_t length_lo;
            uint8_t value[0];
        };
```

I think changing the type from string to bytes would push some slight complexity to consumers of this metadata, which doesn't seem unreasonable.

File: source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc [hunk 12]
```diff
@@ -186,46 +188,21 @@ std::unordered_map<uint8_t, std::vector<unsigned char>> UpstreamProxyProtocolSoc
...
-      continue;
-    }
-    std::vector<unsigned char> value_bytes(sanitised_value.begin(), sanitised_value.end());
-    custom_tlvs[tlv_type] = value_bytes;
+  envoy::extensions::transport_sockets::proxy_protocol::v3::CustomTlvMetadata custom_metadata;
+  if (!filter_it->second.UnpackTo(&custom_metadata)) {
+    ENVOY_LOG(warn, "failed to unpack custom TLVs from upstream host metadata");
+    return custom_tlvs;
+  }
+  // TODO: Prevent empty values from being added to the map? Is that possible with protobuf validation?
```
Comment (timflannagan) [inline source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc]: TODO: I need to test this behavior with a custom build.

Comment (timflannagan) [reply source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc]: Hmm, the protobuf length validation for that `value` key didn't seem to work when I deployed a custom build and providing a custom TLV type with an empty string value. @yuval-k I'm less familiar with how protobuf validation is handled in envoy, but we can fallback to handling this edge case in code too.

File: api/envoy/extensions/transport_sockets/proxy_protocol/v3/upstream_proxy_protocol.proto (hunk 11 above)
Comment (yuval-k) [reply api/envoy/extensions/transport_sockets/proxy_protocol/v3/upstream_proxy_protocol.proto]: I think it makes sense to change the value to bytes.

File: source/extensions/common/proxy_protocol/proxy_protocol_header.cc (hunk 2 above)
Comment (yuval-k) [reply source/extensions/common/proxy_protocol/proxy_protocol_header.cc]: i would have worded my comment better; as the number of items here will likely be less than 10, i don't think it matters if we use map or vector.

File: source/extensions/common/proxy_protocol/proxy_protocol_header.cc (hunk 10 above)
Comment (yuval-k) [reply source/extensions/common/proxy_protocol/proxy_protocol_header.cc:118]: i was thinking that a user might have the same type twice in the protobuf; but that's a control-plane bug. so i'm ok with this as is

Comment (timflannagan) [reply source/extensions/common/proxy_protocol/proxy_protocol_header.cc:118]: Ah I see what you mean. We handled that edge case in the previous implementation by skipping duplicate keys. Agreed it would be a control plane bug, but I think we can still be handle those edge case in the buildCustomTLVs code.

Comment (timflannagan) [reply source/extensions/common/proxy_protocol/proxy_protocol_header.cc:118]: Handled those edge cases in https://github.com/timflannagan/envoy/pull/1/commits/8560b3dec09b54945f654a9a4c9dcb020611630d and added a unit test in https://github.com/timflannagan/envoy/pull/1/commits/26217fd1d743dcceb3409e430affc0fd32afa387.
